File: implementations/rust/src/bitvector.rs
//! Fixed-length bit vector implementation using 32-bit words.
//!
//! This module provides fixed-length bit vector operations optimized for
//! POCKET+ compression. Uses 32-bit words with big-endian byte packing to
//! match ESA/ESOC reference implementation.
//!
//! ## Bit Numbering Convention (CCSDS 124.0-B-1 Section 1.6.1)
//! - Bit 0 = LSB (Least Significant Bit)
//! - Bit N-1 = MSB (Most Significant Bit, transmitted first)
//!
//! ## Word Packing (Big-Endian)
//! Within each 32-bit word:
//! - Word\[i\] = (Byte\[4i\] << 24) | (Byte\[4i+1\] << 16) | (Byte\[4i+2\] << 8) | Byte\[4i+3\]
//! - Bit 0 = LSB of word, Bit 31 = MSB of word

#![allow(clippy::cast_possible_truncation)]
#![allow(clippy::cast_sign_loss)]
#![allow(clippy::return_self_not_must_use)]

/// Maximum packet length in bits (CCSDS max).
pub const MAX_PACKET_LENGTH: usize = 65535;

/// Fixed-length bit vector structure.
///
/// Stores a binary vector of length F bits using 32-bit words.
/// Bit 0 is the LSB, bit F-1 is the MSB.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct BitVector {
    /// 32-bit word storage (big-endian packing).
    data: Vec<u32>,
    /// Number of bits (F).
    length: usize,
}

impl BitVector {
    /// Create a new bit vector with specified length, initialized to zero.
    ///
    /// # Arguments
    /// * `num_bits` - Number of bits (1 to `MAX_PACKET_LENGTH`)
    ///
    /// # Returns
    /// A new `BitVector` with all bits set to zero.
    ///
    /// # Panics
    /// Panics if `num_bits` is 0 or exceeds `MAX_PACKET_LENGTH`.
    pub fn new(num_bits: usize) -> Self {
        assert!(num_bits > 0 && num_bits <= MAX_PACKET_LENGTH);

        // Calculate number of 32-bit words needed
        let num_bytes = (num_bits + 7) / 8;
        let num_words = (num_bytes + 3) / 4; // Ceiling division

        Self {
            data: vec![0u32; num_words],
            length: num_bits,
        }
    }

    /// Create a bit vector from raw bytes.
    ///
    /// # Arguments
    /// * `bytes` - Input byte slice
    /// * `num_bits` - Number of valid bits
    ///
    /// # Panics
    /// Panics if byte slice is too short for the specified bit count.
    pub fn from_bytes(bytes: &[u8], num_bits: usize) -> Self {
        assert!(num_bits > 0 && num_bits <= MAX_PACKET_LENGTH);

        let expected_bytes = (num_bits + 7) / 8;
        assert!(bytes.len() >= expected_bytes);

        let num_words = (expected_bytes + 3) / 4;
        let mut data = vec![0u32; num_words];

        // Pack bytes into 32-bit words (big-endian)
        let mut j = 4u32; // Counter for bytes within word (4, 3, 2, 1)
        let mut bytes_to_int = 0u32;
        let mut current_word = 0usize;

        for &byte in bytes.iter().take(expected_bytes) {
            j -= 1;
            bytes_to_int |= u32::from(byte) << (j * 8);

            if j == 0 {
                // Word complete - store it
                data[current_word] = bytes_to_int;
                current_word += 1;
                bytes_to_int = 0;
                j = 4;
            }
        }

        // Handle incomplete final word
        if j < 4 {
            data[current_word] = bytes_to_int;
        }

        Self {
            data,
            length: num_bits,
        }
    }

    /// Convert bit vector to bytes.
    ///
    /// # Returns
    /// A new `Vec<u8>` containing the bit vector data.
    pub fn to_bytes(&self) -> Vec<u8> {
        let expected_bytes = (self.length + 7) / 8;
        let mut result = Vec::with_capacity(expected_bytes);

        let mut byte_index = 0usize;
        for word in &self.data {
            // Extract up to 4 bytes from this word (big-endian)
            for j in (0u32..4).rev() {
                if byte_index >= expected_bytes {
                    break;
                }
                result.push((word >> (j * 8)) as u8);
                byte_index += 1;
            }
        }

        result
    }

    /// Set all bits to zero.
    pub fn zero(&mut self) {
        for word in &mut self.data {
            *word = 0;
        }
    }

    /// Get the length in bits.
    #[inline]
    pub fn len(&self) -> usize {
        self.length
    }

    /// Check if the bit vector is empty (zero length).
    #[inline]
    pub fn is_empty(&self) -> bool {
        self.length == 0
    }

    /// Get raw access to the underlying 32-bit words.
    ///
    /// Used by encoding functions for efficient word-level operations.
    #[inline]
    pub fn words(&self) -> &[u32] {
        &self.data
    }

    /// Get bit value at position.
    ///
    /// # Arguments
    /// * `pos` - Bit position (0 = LSB, length-1 = MSB)
    ///
    /// # Returns
    /// Bit value (0 or 1), or 0 if position is out of bounds.
    #[inline]
    pub fn get_bit(&self, pos: usize) -> u8 {
        if pos >= self.length {
            return 0;
        }

        // Direct bit-to-word mapping (optimized):
        // word_index = pos / 32, bit_in_word = 31 - (pos % 32)
        // MSB-first: bit 0 is at position 31 in word 0
        let word_index = pos >> 5;
        let bit_in_word = 31 - (pos & 31);

        ((self.data[word_index] >> bit_in_word) & 1) as u8
    }

    /// Set bit value at position.
    ///
    /// # Arguments
    /// * `pos` - Bit position (0 = LSB, length-1 = MSB)
    /// * `value` - Bit value (0 or non-zero for 1)
    #[inline]
    pub fn set_bit(&mut self, pos: usize, value: u8) {
        if pos >= self.length {
            return;
        }

        // Direct bit-to-word mapping (optimized):
        // word_index = pos / 32, bit_in_word = 31 - (pos % 32)
        // MSB-first: bit 0 is at position 31 in word 0
        let word_index = pos >> 5;
        let bit_in_word = 31 - (pos & 31);

        if value != 0 {
            self.data[word_index] |= 1 << bit_in_word;
        } else {
            self.data[word_index] &= !(1 << bit_in_word);
        }
    }

    /// Bitwise XOR of two bit vectors.
    ///
    /// # Arguments
    /// * `other` - Other bit vector (must have same length)
    ///
    /// # Returns
    /// A new `BitVector` containing the XOR result.
    pub fn xor(&self, other: &Self) -> Self {
        let num_words = self.data.len().min(other.data.len());
        let mut result = Self::new(self.length);

        for i in 0..num_words {
            result.data[i] = self.data[i] ^ other.data[i];
        }

        result
    }

    /// Bitwise OR of two bit vectors.
    ///
    /// # Arguments
    /// * `other` - Other bit vector (must have same length)
    ///
    /// # Returns
    /// A new `BitVector` containing the OR result.
    pub fn or(&self, other: &Self) -> Self {
        let num_words = self.data.len().min(other.data.len());
        let mut result = Self::new(self.length);

        for i in 0..num_words {
            result.data[i] = self.data[i] | other.data[i];
        }

        result
    }

    /// In-place bitwise OR with another bit vector.
    ///
    /// # Arguments
    /// * `other` - Other bit vector (must have same length)
    #[inline]
    pub fn or_assign(&mut self, other: &Self) {
        let num_words = self.data.len().min(other.data.len());
        for i in 0..num_words {
            self.data[i] |= other.data[i];
        }
    }

    /// Bitwise AND of two bit vectors.
    ///
    /// # Arguments
    /// * `other` - Other bit vector (must have same length)
    ///
    /// # Returns
    /// A new `BitVector` containing the AND result.
    pub fn and(&self, other: &Self) -> Self {
        let num_words = self.data.len().min(other.data.len());
        let mut result = Self::new(self.length);

        for i in 0..num_words {
            result.data[i] = self.data[i] & other.data[i];
        }

        result
    }

    /// Bitwise NOT of the bit vector.
    ///
    /// # Returns
    /// A new `BitVector` containing the NOT result.
    pub fn not(&self) -> Self {
        let mut result = Self::new(self.length);

        for i in 0..self.data.len() {
            result.data[i] = !self.data[i];
        }

        // Mask off unused bits in last word with big-endian packing
        if !self.data.is_empty() {
            let num_bytes = (self.length + 7) / 8;
            let bytes_in_last_word = ((num_bytes - 1) % 4) + 1;
            let bits_in_last_byte = self.length - ((num_bytes - 1) * 8);

            // Create mask for valid bits in big-endian word
            let mut mask = 0u32;
            for byte in 0..bytes_in_last_word {
                let byte_mask: u8 = if byte == bytes_in_last_word - 1 {
                    // Handle case where bits_in_last_byte is 8 (full byte)
                    if bits_in_last_byte >= 8 {
                        0xFF
                    } else {
                        ((1u32 << bits_in_last_byte) - 1) as u8
                    }
                } else {
                    0xFF
                };
                let shift_amt = (3 - byte as u32) * 8;
                mask |= u32::from(byte_mask) << shift_amt;
            }
            result.data[self.data.len() - 1] &= mask;
        }

        result
    }

    /// Left shift the bit vector by 1 position.
    ///
    /// MSB-first with big-endian word packing:
    /// Left shift means shift towards MSB (bit 0).
    ///
    /// # Returns
    /// A new `BitVector` containing the shifted result.
    pub fn left_shift(&self) -> Self {
        let mut result = Self::new(self.length);

        if !self.data.is_empty() {
            // Process words from first (MSB) to last (LSB)
            for i in 0..self.data.len() - 1 {
                // Shift current word left by 1, bring in MSB from next word
                result.data[i] = (self.data[i] << 1) | (self.data[i + 1] >> 31);
            }
            // Last word: shift left, LSB becomes 0
            result.data[self.data.len() - 1] = self.data[self.data.len() - 1] << 1;
        }

        result
    }

    /// Calculate the Hamming weight (number of 1 bits).
    ///
    /// # Returns
    /// Count of bits set to 1.
    pub fn hamming_weight(&self) -> usize {
        let mut count = 0usize;

        // Count '1' bits in each word using popcount
        for &word in &self.data {
            count += word.count_ones() as usize;
        }

        // Adjust for any extra bits in last word
        let num_bytes = (self.length + 7) / 8;
        let extra_bits = (num_bytes * 8) - self.length;
        if extra_bits > 0 && !self.data.is_empty() {
            // Count bits in the unused portion of the last word and subtract
            let last_word = self.data[self.data.len() - 1];
            let mask = (1u32 << extra_bits) - 1; // Mask for the unused LSBs
            let extra_word = last_word & mask;
            count -= extra_word.count_ones() as usize;
        }

        count
    }

    /// Reverse the bit order.
    ///
    /// # Returns
    /// A new `BitVector` with reversed bit order.
    pub fn reverse(&self) -> Self {
        let mut result = Self::new(self.length);

        for i in 0..self.length {
            let bit = self.get_bit(i);
            result.set_bit(self.length - 1 - i, bit);
        }

        result
    }

    /// Copy from another bit vector.
    ///
    /// # Arguments
    /// * `other` - Source bit vector
    #[inline]
    pub fn copy_from(&mut self, other: &Self) {
        if self.data.len() == other.data.len() {
            // Fast path: same size, direct copy
            self.data.copy_from_slice(&other.data);
        } else {
            // Slow path: resize needed
            self.data.resize(other.data.len(), 0);
            self.data.copy_from_slice(&other.data);
        }
        self.length = other.length;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new() {
        let bv = BitVector::new(720);
        assert_eq!(bv.len(), 720);
        assert_eq!(bv.hamming_weight(), 0);
    }

    #[test]
    fn test_get_set_bit() {
        let mut bv = BitVector::new(32);

        // Set bit 0 (LSB)
        bv.set_bit(0, 1);
        assert_eq!(bv.get_bit(0), 1);
        assert_eq!(bv.get_bit(1), 0);

        // Set bit 31 (MSB for 32-bit)
        bv.set_bit(31, 1);
        assert_eq!(bv.get_bit(31), 1);

        // Clear bit 0
        bv.set_bit(0, 0);
        assert_eq!(bv.get_bit(0), 0);
    }

    #[test]
    fn test_from_bytes_to_bytes_roundtrip() {
        let original = vec![0xDE, 0xAD, 0xBE, 0xEF, 0xCA, 0xFE];
        let bv = BitVector::from_bytes(&original, 48);
        let result = bv.to_bytes();
        assert_eq!(result, original);
    }

    #[test]
    fn test_xor() {
        let mut a = BitVector::new(32);
        let mut b = BitVector::new(32);

        a.set_bit(0, 1);
        a.set_bit(1, 1);
        b.set_bit(1, 1);
        b.set_bit(2, 1);

        let result = a.xor(&b);
        assert_eq!(result.get_bit(0), 1); // 1 XOR 0 = 1
        assert_eq!(result.get_bit(1), 0); // 1 XOR 1 = 0
        assert_eq!(result.get_bit(2), 1); // 0 XOR 1 = 1
    }

    #[test]
    fn test_or() {
        let mut a = BitVector::new(32);
        let mut b = BitVector::new(32);

        a.set_bit(0, 1);
        b.set_bit(1, 1);

        let result = a.or(&b);
        assert_eq!(result.get_bit(0), 1);
        assert_eq!(result.get_bit(1), 1);
        assert_eq!(result.get_bit(2), 0);
    }

    #[test]
    fn test_and() {
        let mut a = BitVector::new(32);
        let mut b = BitVector::new(32);

        a.set_bit(0, 1);
        a.set_bit(1, 1);
        b.set_bit(1, 1);
        b.set_bit(2, 1);

        let result = a.and(&b);
        assert_eq!(result.get_bit(0), 0); // 1 AND 0 = 0
        assert_eq!(result.get_bit(1), 1); // 1 AND 1 = 1
        assert_eq!(result.get_bit(2), 0); // 0 AND 1 = 0
    }

    #[test]
    fn test_not() {
        let mut bv = BitVector::new(8);
        bv.set_bit(0, 1);
        bv.set_bit(2, 1);

        let result = bv.not();
        assert_eq!(result.get_bit(0), 0);
        assert_eq!(result.get_bit(1), 1);
        assert_eq!(result.get_bit(2), 0);
        assert_eq!(result.get_bit(3), 1);
    }

    #[test]
    fn test_left_shift() {
        let mut bv = BitVector::new(32);
        bv.set_bit(1, 1); // Set bit 1

        let result = bv.left_shift();
        assert_eq!(result.get_bit(0), 1); // Bit 1 shifted to bit 0
        assert_eq!(result.get_bit(1), 0);
    }

    #[test]
    fn test_hamming_weight() {
        let mut bv = BitVector::new(32);
        assert_eq!(bv.hamming_weight(), 0);

        bv.set_bit(0, 1);
        bv.set_bit(5, 1);
        bv.set_bit(31, 1);
        assert_eq!(bv.hamming_weight(), 3);
    }

    #[test]
    fn test_zero() {
        let mut bv = BitVector::new(32);
        bv.set_bit(0, 1);
        bv.set_bit(15, 1);
        bv.set_bit(31, 1);

        bv.zero();
        assert_eq!(bv.hamming_weight(), 0);
    }

    #[test]
    fn test_equals() {
        let mut a = BitVector::new(32);
        let mut b = BitVector::new(32);

        assert_eq!(a, b);

        a.set_bit(5, 1);
        assert_ne!(a, b);

        b.set_bit(5, 1);
        assert_eq!(a, b);
    }

    #[test]
    fn test_720_bits() {
        // Test with POCKET+ standard packet size
        let mut bv = BitVector::new(720);
        assert_eq!(bv.len(), 720);

        // Set first and last bits
        bv.set_bit(0, 1);
        bv.set_bit(719, 1);
        assert_eq!(bv.get_bit(0), 1);
        assert_eq!(bv.get_bit(719), 1);
        assert_eq!(bv.hamming_weight(), 2);

        // Round-trip through bytes
        let bytes = bv.to_bytes();
        assert_eq!(bytes.len(), 90); // 720 / 8 = 90 bytes

        let bv2 = BitVector::from_bytes(&bytes, 720);
        assert_eq!(bv, bv2);
    }

    #[test]
    fn test_reverse() {
        let mut bv = BitVector::new(8);
        bv.set_bit(0, 1); // bit 0
        bv.set_bit(1, 0);
        bv.set_bit(2, 1); // bit 2

        let rev = bv.reverse();
        assert_eq!(rev.get_bit(7), 1); // bit 0 -> bit 7
        assert_eq!(rev.get_bit(6), 0);
        assert_eq!(rev.get_bit(5), 1); // bit 2 -> bit 5
    }

    #[test]
    fn test_default() {
        let bv = BitVector::default();
        assert_eq!(bv.len(), 0);
        assert!(bv.is_empty());
    }
}

File: implementations/java/src/main/java/space/tanagra/pocketplus/BitVector.java
/*
 * Copyright (c) 2025 Tanagra Space
 * SPDX-License-Identifier: MIT
 */

package space.tanagra.pocketplus;

import java.util.Arrays;

/**
 * Fixed-length bit vector with 32-bit word storage.
 *
 * <p>This class provides efficient bit manipulation operations for the POCKET+ algorithm. Uses
 * MSB-first bit ordering matching the ESA/ESOC reference implementation.
 *
 * <p>Bit numbering (CCSDS 124.0-B-1 Section 1.6.1):
 *
 * <ul>
 *   <li>Bit 0 = first transmitted bit (MSB of first byte)
 *   <li>Bit N-1 = last transmitted bit
 * </ul>
 *
 * <p>Word packing (big-endian):
 *
 * <ul>
 *   <li>Word[i] = (Byte[4i] &lt;&lt; 24) | (Byte[4i+1] &lt;&lt; 16) | (Byte[4i+2] &lt;&lt; 8) |
 *       Byte[4i+3]
 *   <li>Bit 0 at position 31 of word 0, Bit 31 at position 0 of word 0
 * </ul>
 */
public final class BitVector {

  /** 32-bit word storage. */
  private final int[] data;

  /** Length in bits. */
  private final int length;

  /** Number of 32-bit words used. */
  private final int numWords;

  /**
   * Creates a new BitVector with the specified number of bits, initialized to zero.
   *
   * @param numBits the number of bits in this vector
   * @throws IllegalArgumentException if numBits is negative or exceeds maximum
   */
  public BitVector(int numBits) {
    if (numBits < 0 || numBits > PocketPlus.MAX_PACKET_LENGTH) {
      throw new IllegalArgumentException("Invalid bit count: " + numBits);
    }
    this.length = numBits;
    int numBytes = (numBits + 7) / 8;
    this.numWords = (numBytes + 3) / 4; // Ceiling division
    this.data = new int[numWords];
  }

  /**
   * Creates a BitVector by copying another BitVector.
   *
   * @param other the BitVector to copy
   */
  public BitVector(BitVector other) {
    this.length = other.length;
    this.numWords = other.numWords;
    this.data = Arrays.copyOf(other.data, other.data.length);
  }

  /**
   * Returns the length of this vector in bits.
   *
   * @return length in bits
   */
  public int length() {
    return length;
  }

  /**
   * Returns the number of 32-bit words.
   *
   * @return number of words
   */
  public int numWords() {
    return numWords;
  }

  /** Sets all bits to zero. */
  public void zero() {
    Arrays.fill(data, 0);
  }

  /**
   * Gets the bit at the specified position.
   *
   * <p>MSB-first: bit 0 is at position 31 in word 0.
   *
   * @param pos bit position (0 = first transmitted bit)
   * @return 0 or 1
   */
  public int getBit(int pos) {
    if (pos < 0 || pos >= length) {
      return 0;
    }
    // Direct bit-to-word mapping (matches C implementation)
    // word_index = pos / 32, bit_in_word = 31 - (pos % 32)
    int wordIndex = pos >>> 5;
    int bitInWord = 31 - (pos & 31);
    return (data[wordIndex] >>> bitInWord) & 1;
  }

  /**
   * Sets the bit at the specified position.
   *
   * <p>MSB-first: bit 0 is at position 31 in word 0.
   *
   * @param pos bit position (0 = first transmitted bit)
   * @param value 0 or 1
   */
  public void setBit(int pos, int value) {
    if (pos < 0 || pos >= length) {
      return;
    }
    // Direct bit-to-word mapping (matches C implementation)
    int wordIndex = pos >>> 5;
    int bitInWord = 31 - (pos & 31);
    if ((value & 1) != 0) {
      data[wordIndex] |= (1 << bitInWord);
    } else {
      data[wordIndex] &= ~(1 << bitInWord);
    }
  }

  /**
   * Returns a new BitVector that is the XOR of this and another vector.
   *
   * @param other the other BitVector
   * @return new BitVector with result
   */
  public BitVector xor(BitVector other) {
    BitVector result = new BitVector(length);
    int words = Math.min(numWords, other.numWords);
    for (int i = 0; i < words; i++) {
      result.data[i] = this.data[i] ^ other.data[i];
    }
    return result;
  }

  /**
   * Returns a new BitVector that is the OR of this and another vector.
   *
   * @param other the other BitVector
   * @return new BitVector with result
   */
  public BitVector or(BitVector other) {
    BitVector result = new BitVector(length);
    int words = Math.min(numWords, other.numWords);
    for (int i = 0; i < words; i++) {
      result.data[i] = this.data[i] | other.data[i];
    }
    return result;
  }

  /**
   * Returns a new BitVector that is the AND of this and another vector.
   *
   * @param other the other BitVector
   * @return new BitVector with result
   */
  public BitVector and(BitVector other) {
    BitVector result = new BitVector(length);
    int words = Math.min(numWords, other.numWords);
    for (int i = 0; i < words; i++) {
      result.data[i] = this.data[i] & other.data[i];
    }
    return result;
  }

  /**
   * Returns a new BitVector that is the bitwise NOT of this vector.
   *
   * @return new BitVector with inverted bits
   */
  public BitVector not() {
    BitVector result = new BitVector(length);
    for (int i = 0; i < numWords; i++) {
      result.data[i] = ~this.data[i];
    }
    // Mask off unused bits in last word with big-endian packing
    if (numWords > 0) {
      int numBytes = (length + 7) / 8;
      int bytesInLastWord = ((numBytes - 1) % 4) + 1;
      int bitsInLastByte = length - ((numBytes - 1) * 8);

      // Create mask for valid bits in big-endian word
      int mask = 0;
      for (int b = 0; b < bytesInLastWord; b++) {
        int byteMask;
        if (b == bytesInLastWord - 1) {
          byteMask = (1 << bitsInLastByte) - 1;
        } else {
          byteMask = 0xFF;
        }
        int shiftAmt = (3 - b) * 8;
        mask |= (byteMask << shiftAmt);
      }
      result.data[numWords - 1] &= mask;
    }
    return result;
  }

  /**
   * Returns a new BitVector that is this vector shifted left by one bit.
   *
   * <p>MSB-first with big-endian word packing: Left shift means shift towards MSB (bit 0). Word
   * level: shift each word left by 1, carry high bit from next word.
   *
   * @return new BitVector shifted left
   */
  public BitVector leftShift() {
    BitVector result = new BitVector(length);
    if (numWords > 0) {
      // Process words from first (MSB) to last (LSB)
      for (int i = 0; i < numWords - 1; i++) {
        // Shift current word left by 1, bring in MSB from next word
        result.data[i] = (data[i] << 1) | (data[i + 1] >>> 31);
      }
      // Last word: shift left, LSB becomes 0
      result.data[numWords - 1] = data[numWords - 1] << 1;
    }
    return result;
  }

  /**
   * Returns a new BitVector with bits reversed.
   *
   * @return new BitVector with reversed bits
   */
  public BitVector reverse() {
    BitVector result = new BitVector(length);
    for (int i = 0; i < length; i++) {
      result.setBit(length - 1 - i, getBit(i));
    }
    return result;
  }

  /**
   * Returns the Hamming weight (number of 1 bits).
   *
   * @return count of set bits
   */
  public int hammingWeight() {
    int count = 0;
    for (int i = 0; i < numWords; i++) {
      count += Integer.bitCount(data[i]);
    }
    // Adjust for extra bits in last word
    int numBytes = (length + 7) / 8;
    int extraBits = (numBytes * 8) - length;
    if (extraBits > 0 && numWords > 0) {
      // Count bits in unused portion and subtract
      int lastWord = data[numWords - 1];
      int mask = (1 << extraBits) - 1;
      count -= Integer.bitCount(lastWord & mask);
    }
    return count;
  }

  /**
   * Checks if all bits are zero.
   *
   * @return true if all bits are zero
   */
  public boolean isZero() {
    for (int i = 0; i < numWords; i++) {
      if (data[i] != 0) {
        return false;
      }
    }
    return true;
  }

  /**
   * Creates a BitVector from a byte array (big-endian packing).
   *
   * @param bytes the byte array
   * @param numBits number of bits to use
   * @return new BitVector
   */
  public static BitVector fromBytes(byte[] bytes, int numBits) {
    BitVector result = new BitVector(numBits);
    int numBytes = (numBits + 7) / 8;

    // Pack bytes into 32-bit words (big-endian)
    int j = 4; // Counter for bytes within word (4, 3, 2, 1)
    int bytesToInt = 0;
    int currentWord = 0;

    for (int i = 0; i < numBytes && i < bytes.length; i++) {
      j--;
      bytesToInt |= (bytes[i] & 0xFF) << (j * 8);

      if (j == 0) {
        // Word complete - store it
        result.data[currentWord] = bytesToInt;
        currentWord++;
        bytesToInt = 0;
        j = 4;
      }
    }

    // Handle incomplete final word
    if (j < 4) {
      result.data[currentWord] = bytesToInt;
    }

    return result;
  }

  /**
   * Converts this BitVector to a byte array (big-endian packing).
   *
   * @return byte array representation
   */
  public byte[] toBytes() {
    int numBytes = (length + 7) / 8;
    byte[] result = new byte[numBytes];

    // Extract bytes from 32-bit words (big-endian)
    int byteIndex = 0;
    for (int wordIndex = 0; wordIndex < numWords && byteIndex < numBytes; wordIndex++) {
      int word = data[wordIndex];

      // Extract up to 4 bytes from this word
      for (int j = 3; j >= 0 && byteIndex < numBytes; j--) {
        result[byteIndex] = (byte) ((word >>> (j * 8)) & 0xFF);
        byteIndex++;
      }
    }

    return result;
  }

  /**
   * Checks equality with another BitVector.
   *
   * @param other the other BitVector
   * @return true if equal
   */
  public boolean equals(BitVector other) {
    if (other == null || this.length != other.length) {
      return false;
    }
    return Arrays.equals(this.data, other.data);
  }

  @Override
  public int hashCode() {
    return Arrays.hashCode(data) ^ length;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj instanceof BitVector) {
      return equals((BitVector) obj);
    }
    return false;
  }

  /**
   * Returns a string representation of this vector (for debugging).
   *
   * @return binary string representation
   */
  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder(length);
    for (int i = 0; i < length; i++) {
      sb.append(getBit(i));
    }
    return sb.toString();
  }

  /**
   * Returns the internal word array (for internal use).
   *
   * @return the data array
   */
  int[] getData() {
    return data;
  }

  /**
   * Gets a word from the internal data array.
   *
   * @param index word index
   * @return word value
   */
  int getWord(int index) {
    if (index >= 0 && index < numWords) {
      return data[index];
    }
    return 0;
  }

  /**
   * Sets a word in the internal data array.
   *
   * @param index word index
   * @param value word value
   */
  void setWord(int index, int value) {
    if (index >= 0 && index < numWords) {
      data[index] = value;
    }
  }

  // ============================================================
  // In-place mutation methods (for performance-critical paths)
  // ============================================================

  /**
   * XORs this vector with another in-place.
   *
   * @param other the other BitVector
   */
  public void xorInPlace(BitVector other) {
    int words = Math.min(numWords, other.numWords);
    for (int i = 0; i < words; i++) {
      data[i] ^= other.data[i];
    }
  }

  /**
   * ORs this vector with another in-place.
   *
   * @param other the other BitVector
   */
  public void orInPlace(BitVector other) {
    int words = Math.min(numWords, other.numWords);
    for (int i = 0; i < words; i++) {
      data[i] |= other.data[i];
    }
  }

  /**
   * ANDs this vector with another in-place.
   *
   * @param other the other BitVector
   */
  public void andInPlace(BitVector other) {
    int words = Math.min(numWords, other.numWords);
    for (int i = 0; i < words; i++) {
      data[i] &= other.data[i];
    }
  }

  /**
   * Copies data from another BitVector into this one.
   *
   * @param other the source BitVector
   */
  public void copyFrom(BitVector other) {
    int words = Math.min(numWords, other.numWords);
    for (int i = 0; i < words; i++) {
      data[i] = other.data[i];
    }
    // Zero remaining words if this is larger
    for (int i = words; i < numWords; i++) {
      data[i] = 0;
    }
  }

  /**
   * Left shifts this vector by one bit in-place.
   *
   * <p>MSB-first: shift towards bit 0.
   */
  public void leftShiftInPlace() {
    if (numWords > 0) {
      for (int i = 0; i < numWords - 1; i++) {
        data[i] = (data[i] << 1) | (data[i + 1] >>> 31);
      }
      data[numWords - 1] = data[numWords - 1] << 1;
    }
  }
}
